x/exchange/client/cli/flags.go

```diff
@@ -0,0 +1,684 @@
+		return def, err
+	}
+	return ParseFeeRatios(vals)
+}
+
+// ParseFeeRatios parses a FeeRatio from each of the provided vals.
+func ParseFeeRatios(vals []string) ([]exchange.FeeRatio, error) {
+	var errs []error
+	ratios := make([]exchange.FeeRatio, 0, len(vals))
+	for _, val := range vals {
+		ratio, err := exchange.ParseFeeRatio(val)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if ratio != nil {
+			ratios = append(ratios, *ratio)
+		}
+	}
+	return ratios, errors.Join(errs...)
+}
+
+// ReadSplitsFlag reads a StringSlice flag and converts it into a slice of exchange.DenomSplit.
+// This assumes that the flag was defined with a default of nil or []string{}.
+func ReadSplitsFlag(flagSet *pflag.FlagSet, name string) ([]exchange.DenomSplit, error) {
+	vals, err := flagSet.GetStringSlice(name)
+	if len(vals) == 0 || err != nil {
+		return nil, err
+	}
+	return ParseSplits(vals)
+}
+
+// ParseSplit parses a DenomSplit from a string with the format "<denom>:<amount>".
+func ParseSplit(val string) (*exchange.DenomSplit, error) {
+	parts := strings.Split(val, ":")
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid denom split %q: expected format <denom>:<amount>", val)
+	}
+
+	denom := strings.TrimSpace(parts[0])
+	amountStr := strings.TrimSpace(parts[1])
+	if len(denom) == 0 || len(amountStr) == 0 {
+		return nil, fmt.Errorf("invalid denom split %q: both a <denom> and <amount> are required", val)
+	}
+
+	amount, err := strconv.ParseUint(amountStr, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse %q amount: %w", val, err)
+	}
+
+	return &exchange.DenomSplit{Denom: denom, Split: uint32(amount)}, nil
+}
+
+// ParseSplits parses a DenomSplit from each of the provided vals.
+func ParseSplits(vals []string) ([]exchange.DenomSplit, error) {
+	var errs []error
+	splits := make([]exchange.DenomSplit, 0, len(vals))
+	for _, val := range vals {
+		split, err := ParseSplit(val)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if split != nil {
+			splits = append(splits, *split)
+		}
+	}
+	return splits, errors.Join(errs...)
+}
+
+// ReadStringFlagOrArg gets a required string from either a flag or the first provided arg.
+// This assumes that the flag was defined with a default of "".
+func ReadStringFlagOrArg(flagSet *pflag.FlagSet, args []string, flagName, varName string) (string, error) {
+	rv, err := flagSet.GetString(flagName)
+	if err != nil {
+		return "", err
+	}
+
+	if len(args) > 0 && len(args[0]) > 0 {
+		if len(rv) > 0 {
+			return "", fmt.Errorf("cannot provide <%s> as both an arg (%q) and flag (--%s %q)", varName, args[0], flagName, rv)
+		}
+
+		return args[0], nil
+	}
+
+	if len(rv) == 0 {
+		return "", fmt.Errorf("no <%s> provided", varName)
+	}
+
+	return rv, nil
+}
+
+// ReadProposalFlag gets the --proposal string value and attempts to read the file in as a Tx in json.
+// It then attempts to extract any messages contained in any govv1.MsgSubmitProposal messages in that Tx.
```

Comment:
Typo govv1

Reply:
I read this as gov-v1... which is what I expected.

Reply:
`govv1` is correct. At the top, we import `govv1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"`. The reason to include the `v1` on that is because there's also `v1beta1` which is older stuff that I think is mostly deprecated.